370-auth-ui/web-ui.md

```diff
@@ -0,0 +1,171 @@
+# Web User Interface
+
+## Overview
+
+Skygear Auth is an OpenID Connect Provider (OP). The Relying Party redirects the user to the authorization endpoint, where the user can sign up, sign in, perform multi-factor authentication, reset forgotten password and authorize access. At the settings endpoint the user can manage their account.
+
+## Page Templates
+
+All page templates are rendered with Go standard library package [html/template](https://golang.org/pkg/html/template/). This is also the package being used to render HTML email template. The developer need not to learn anything new.
+
+## Traditional form submission and JavaScript
+
+The interaction between Skygear Auth and the user is traditional form submission. JavaScript is only required to add extra functionalities such as dynamically checking password requirements and toggling of password visibility.
+
+## Customization of page templates
+
+The developer can customize the pages at 2 levels. The can keep using the default templates and only customize the theme. Or the developer can replace the default templates with their own templates.
+
+### Customizing the theme with CSS
+
+The developer can specify a snippet of CSS to be inlined into every pages. The developer is strongly recommended to write CSS covering only documented CSS classes, CSS pseudo classes and CSS properties.
+
+CSS classes:
+
+- `.primary-txt`: The primary text.
+- `.secondary-txt`: The secondary text.
+- `.primary-btn`: The primary button.
+- `.secondary-btn`: The secondary button.
+- `.anchor`: The anchor.
+- `.text-input`: The text input.
+
+CSS pseudo classes:
+
+- `:link`: Unvisited links.
+- `:visited`: Visited links.
+- `:hover`: Mouse hover state.
+- `:active`: Button being clicked.
+
+CSS properties:
+
+- `color`
+- `background-color`
+- `border*`
+
+### Customizing with own templates
+
+The developer can also provide their own templates via [the existing template mechanism](../251-templates/index.md).
+
+## The authorization endpoint
+
+The authorization endpoint `/authorize` is the entry point of OIDC. The user can sign up, sign in, perform multi-factor authentication, reset forgotten password and authorize access.
```

Comment:
We may want to add a version prefix here? e.g. `/v1/authorize`.

Reply:
But the Authorization Server Metadata is a well known static endpoint. The metadata itself does not support versioning.

Reply:
For example, [Google's discovery document](https://accounts.google.com/.well-known/openid-configuration) has a versioned authorization endpoint: `https://accounts.google.com/o/oauth2/v2/auth`

The protocol itself may be relatively stable, but the behavior may have breaking changes, so adding API versioning would be a good practice.

Reply:
I mean the `/.well-known/openid-configuration` endpoint is not versioned at all. Today the `authorization_endpoint` is `https://accounts.google.com/o/oauth2/v2/auth` but tomorrow it may be `https://accounts.google.com/o/oauth3/v3/auth`. However as long as it is a OAuth 2.0 or OpenID Connect 1.0 metadata document, all endpoints should behave according to the spec.

The OIDC client should not interpret the versioning scheme of the endpoint URL. Instead the client look at other metadata to find out what features the AS supports. If the OAuth were 3, I guess the standard committee will introduce another well known endpoint for the AS to advertise its capabilties?

Reply:
Yes, but we cannot just throw away OAuth2 & OIDC when we support OAuth3. Then, we'll have two authorize endpoint for each protocols, and their respective discovery document would refer to individual versioned authorized endpoint: e.g. `/v1/authorize` for OAuth2, `/v2/authorize` for OAuth3.

Reply:
I prefer the prefix `/oauth2/authorize`, what do you think?

Reply:
I think it's good enough for authorization endpoint.